ranger/ranger.go

```diff
@@ -41,69 +37,82 @@ type Ranger struct {
 	ctx       context.Context
 	db        postgres.DatabaseService
 	env       trails.Environment
-	kr        keyring.Keyringable
-	p         template.Parser
-	sessions  session.SessionStorer
+	metadata  Metadata
 	shutdowns []ShutdownFn
 	srv       *http.Server
-	url       *url.URL
-	userstore middleware.UserStorer
 }
 
 // New constructs a Ranger from the provided options.
 // Default options are applied first followed by the options passed into New.
 // Options supplied to New overwrite default configurations.
-func New[User RangerUser](cfg Config[U], opts ...RangerOption) (*Ranger, error) {
+func New[U RangerUser](cfg Config[U]) (*Ranger, error) {
+	err := cfg.Valid()
+	if err != nil {
+		return nil, err
+	}
+
 	r := new(Ranger)
-	followups := make([]OptFollowup, 0)
 
 	// Setup initial configuration
 	r.env = trails.EnvVarOrEnv(environmentEnvVar, trails.Development)
+	r.Logger = defaultLogger()
+	r.ctx, r.cancel = context.WithCancel(context.Background())
 
-	// NOTE(dlk): calling an option configures the *Ranger under construction.
-	// Some options require data from other options.
-	// These options, therefore, must delay configuring the *Ranger
-	// until either (1) user supplied RangerOptions or (2) default RangerOptions
-	// configure the *Ranger first.
-	// They return an optFollowup to be called after the initial set of options are run.
-	for _, opt := range append(defaultOpts(), opts...) {
-		fn, err := opt(r)
-		if err != nil {
-			return r, fmt.Errorf("%w: %s", trails.ErrBadConfig, err)
-		}
+	r.db, err = defaultDB(r.env, cfg.Migrations)
+	if err != nil {
+		return nil, err
+	}
 
-		if fn != nil {
-			followups = append(followups, fn)
-		}
+	r.metadata, err = newMetadata()
+	if err != nil {
+		return nil, err
 	}
 
-	r.userstore = cfg.defaultUserStore(r.db)
+	url := trails.EnvVarOrURL(baseURLEnvVar, defaultBaseURL)
```

Comment:
nit: I'd prefer something like `rootURL` here (and in `defaultResponder`) to be more explicit and to avoid name confusion/collision with `net/url`.